velox/exec/HashTable.h

```diff
@@ -978,6 +1076,8 @@ class HashTable : public BaseHashTable {
   int32_t nextOffset_;
   char** table_ = nullptr;
   memory::ContiguousAllocation tableAllocation_;
+  char** keyAndPointers_ = nullptr;
```

Comment:
`table_` is typed as `char**` for `kArray` usage, I don't see the point to type it the same since you are only handling `kNormalizedKey`

Reply:
@Yuhta Yu Thank you so much for replying so quickly

Reply:
The main ideas of this optimization are as follows:
We noticed that putting tags together can achieve better data locality and improve probe performance. But there may be a problem here. In hits-offten scenarios (when fullProbe is needed), if tags and pointers are not placed in the same cache-line, the speed of accessing keys may become slower. see ：https://github.com/facebookincubator/velox/pull/5637

In order to solve this problem: to obtain better performance in both hits-offten and miss-offten scenarios, I put the key and pointer together, so that in the kNormalizedKey scenario, the pointer does not need to be used when probing, so can reduce a period of random memory access.

Here is the bencmark result, (you could reproduce it by hard code):
`
  bool enableOpt() const {
    return false;
    // return hashMode_ == BaseHashTable::HashMode::kNormalizedKey && isJoinBuild_;
  }
`
return false:

*** Results:
Hit4M: Rows=4000000 Hit%=100 NumProbes=4000000 hash/row=3.40778 probe clocks=99.639 f14Probe=184.447 (185.115%)
 numDistinct=4000000 mode=normalized key
Miss4M: Rows=4000000 Hit%=5 NumProbes=80000000 hash/row=4.36014 probe clocks=82.907 f14Probe=139.33 (168.056%)
 numDistinct=4798469 mode=normalized key
Hit32M: Rows=32000000 Hit%=100 NumProbes=32000000 hash/row=4.29268 probe clocks=145.509 f14Probe=230.778 (158.601%)
 numDistinct=32000000 mode=normalized key
Miss32M: Rows=32000000 Hit%=5 NumProbes=640000000 hash/row=4.31341 probe clocks=115.399 f14Probe=168.091 (145.66%)
 numDistinct=38385441 mode=normalized key

return hashMode_ == BaseHashTable::HashMode::kNormalizedKey && isJoinBuild_;
*** Results:
Hit4M: Rows=4000000 Hit%=100 NumProbes=4000000 hash/row=3.35768 probe clocks=87.662 f14Probe=188.867 (215.449%)
 numDistinct=4000000 mode=normalized key
Miss4M: Rows=4000000 Hit%=5 NumProbes=80000000 hash/row=4.41905 probe clocks=61.0742 f14Probe=135.808 (222.366%)
 numDistinct=4798469 mode=normalized key
Hit32M: Rows=32000000 Hit%=100 NumProbes=32000000 hash/row=4.45841 probe clocks=134.332 f14Probe=243.39 (181.186%)
 numDistinct=32000000 mode=normalized key
Miss32M: Rows=32000000 Hit%=5 NumProbes=640000000 hash/row=4.45149 probe clocks=85.9248 f14Probe=172.104 (200.296%)
 numDistinct=38385441 mode=normalized key


In the case of tags-probe, I think data locality is mainly reflected in TBL-miss. Here is some test data:
opt:
![image](https://github.com/user-attachments/assets/c20a4662-04ab-4c26-b81e-70dd46aac2d1)
base:
![image](https://github.com/user-attachments/assets/e2bdcca5-6fae-47b6-9ae0-107eb4b8a82b)


I guess this is reasonable since the size of tags-array is 1/8 of the original (16bytes vs 128bytes)

Reply:
@Yuhta  @oerling 
Any suggestions for this optimization proposal? Is it worth introducing new data structures into velox?

Reply:
Orri is on PTO this week and will be on VLDB next week.  We need to discuss this after VLDB and see whether and how we want to incorporate this idea.  We used to have same layout but we switched to the current one after some experiments, need to decide the condition in which case the old layout is more efficient.

Reply:
Furthermore, I think tag-probe can be completely replaced by simd-bloom-filter (it can be pushed down by the optimizer). For HashTable, we should assume that hits-offten is the norm.
@Yuhta  @oerling